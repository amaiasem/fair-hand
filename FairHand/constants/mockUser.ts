const mockUser = {

  _id: '604f3dcafcd75136b496ec65',
  name: 'Amaia Semper',
  surname: '',
  email: 'amaia@gmail.com',
  password: 'ahg56',
  image: 'https://instagram.fqsa1-1.fna.fbcdn.net/v/t51.2885-15/e35/s1080x1080/60793471_340043599979046_7591053955213517299_n.jpg?tp=1&_nc_ht=instagram.fqsa1-1.fna.fbcdn.net&_nc_cat=106&_nc_ohc=RbfXgsjm6WgAX_p8uD1&oh=653befd6c914a74146eedf0ff34134cd&oe=607936F6',
  address: 'Travessera de les Corts, Barcelona',
  myFavourites: [
    {
      shopId: '6046ab3fb9f8122550ebebc9'
    }
  ]
}

export default mockUser
