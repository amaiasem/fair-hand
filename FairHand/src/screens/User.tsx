import React from 'react'
import { StyleSheet, Text, View, Image } from 'react-native'
import renderHeader from './Header'
import { COLOR, SIZES, mockUser } from '../../constants'
import { TouchableOpacity } from 'react-native-gesture-handler'

const styles = StyleSheet.create({
  container: {
    marginTop: 30,
    flex: 1
  },
  userInfo: {
    flex: 0.5,
    justifyContent: 'center',
    alignItems: 'center'
  },
  userImage: {
    height: 150,
    width: 150,
    borderRadius: 80,
    margin: 20
  },
  userName: {
    fontSize: SIZES.h1,
    fontWeight: '600'
  },
  userEmail: {
    fontSize: SIZES.h2,
    fontWeight: '600'
  },
  containerButtons: {
    flex: 0.4,
    justifyContent: 'flex-end',
    alignItems: 'center'
  },
  userButtons: {
    width: SIZES.width * 0.85,
    height: 35,
    marginTop: 10,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLOR.orange,
    borderRadius: SIZES.buttonRadius
  },
  userButtonText: {
    color: COLOR.white,
    fontSize: SIZES.buttonText,
    fontWeight: '700'
  },
  logoutButton: {
    width: SIZES.width * 0.85,
    height: 35,
    marginTop: 10,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLOR.black,
    borderRadius: SIZES.buttonRadius
  }

})
// const user = mockUser

const User = ({ mockUser: any }) => {
  return (
    <View style = {styles.container}>
      {renderHeader()}
      <View style={styles.userInfo}>
        <Image style={styles.userImage} source={{ uri: mockUser?.image }}></Image>
        <Text style={styles.userName}>{mockUser?.name}</Text>
        <Text style={styles.userEmail}>{mockUser?.email}</Text>
      </View>
      <View style={styles.containerButtons}>
        <TouchableOpacity style={styles.userButtons}>
          <Text style={styles.userButtonText}>My favourites</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.userButtons}>
          <Text style={styles.userButtonText}>My reviews</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.logoutButton}>
          <Text style={styles.userButtonText}>Logout</Text>
        </TouchableOpacity>
      </View>
    </View>
  )
}

export default User
