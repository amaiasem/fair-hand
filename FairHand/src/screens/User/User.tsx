import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { userUpdate, userLogout } from '../../redux/actions/fairHandActionCreators'
import { StyleSheet, Text, View, Image, Platform } from 'react-native'
import { Ionicons } from '@expo/vector-icons'
import renderHeader from '../../Components/header/Header'
import { COLOR, SIZES, images } from '../../../constants'
import { TouchableOpacity } from 'react-native-gesture-handler'
import UserInterface from '../../Interfaces/userInterface'
import * as ImagePicker from 'expo-image-picker'
import Screen from '../../../constants/ScreenConstants'

const styles = StyleSheet.create({
  container: {
    marginTop: 30,
    flex: 1
  },
  userInfo: {
    flex: 0.5,
    justifyContent: 'center',
    alignItems: 'center'
  },
  userImage: {
    height: 150,
    width: 150,
    borderRadius: 80,
    margin: 20
  },
  userName: {
    fontSize: SIZES.h1,
    fontWeight: '600'
  },
  userEmail: {
    fontSize: SIZES.h2,
    fontWeight: '600'
  },
  camera: {
    position: 'absolute',
    top: '60%',
    right: '30%'
  },
  containerButtons: {
    flex: 0.4,
    justifyContent: 'flex-end',
    alignItems: 'center'
  },
  userButtons: {
    width: SIZES.width * 0.85,
    height: 35,
    marginTop: 10,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLOR.orange,
    borderRadius: SIZES.buttonRadius
  },
  userButtonText: {
    color: COLOR.white,
    fontSize: SIZES.buttonText,
    fontWeight: '700'
  },
  logoutButton: {
    width: SIZES.width * 0.85,
    height: 35,
    marginTop: 10,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: COLOR.black,
    borderRadius: SIZES.buttonRadius
  }

})

const User = ({ user, navigation, action }: {user: UserInterface, navigation: any, action: any}) => {
  const [image, setImage] = React.useState(null)
  const [showButton, setShowButton] = React.useState(false)

  useEffect(() => {
    (async () => {
      if (Platform.OS !== 'web') {
        const { status } = await ImagePicker.requestMediaLibraryPermissionsAsync()
        if (status !== 'granted') {
          alert('Sorry, we need camera roll permissions to make this work!')
        }
      }
    })()
  }, [])

  const pickImage = async () => {
    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.All,
      allowsEditing: true,
      aspect: [1, 1],
      quality: 1
    })
    if (!result.cancelled) {
      setImage(result.uri)
      setShowButton(true)
    }
  }

  function updateImage () {
    const userToUpdate = {
      _id: user._id,
      image: image
    }
    action.userUpdate(userToUpdate)
    setShowButton(false)
  }

  function userLogoutAndRedirect () {
    action.userLogout()
    navigation.navigate(Screen.AppCover)
  }

  return (
    <View style = {styles.container}>
      {renderHeader()}
      <View style={styles.userInfo}>
        <Image style={styles.userImage} source={{ uri: image || user.image || images.userImage }}></Image>
        <Text style={styles.userName}>{user?.name}</Text>
        <Text style={styles.userEmail}>{user?.email}</Text>
        <View style={styles.camera}>
          <Ionicons
          name="camera-outline"
          size={28}
          color={COLOR.orange}
          onPress={pickImage}
          testID='pickImage'/>
        </View>
      </View>
      <View style={styles.containerButtons}>
        {
          showButton
            ? <TouchableOpacity
          disabled={!image}
          style={styles.logoutButton}
          testID='update-image'
          onPress={() => updateImage()}>
            <Text style={styles.userButtonText}>Update image</Text>
          </TouchableOpacity>
            : <Text></Text>
        }

        <TouchableOpacity
        style={styles.userButtons}
        onPress={() => navigation.navigate(Screen.MyFavourites)}
        testID='navigate-myfavourites'>
          <Text style={styles.userButtonText}>My favourites</Text>
        </TouchableOpacity>
        <TouchableOpacity
        style={styles.userButtons}
        testID='navigate-myreviews'
        onPress={() => navigation.navigate(Screen.MyReviews)}>
          <Text style={styles.userButtonText}>My reviews</Text>
        </TouchableOpacity>
        <TouchableOpacity
        style={styles.logoutButton}
        onPress={() => userLogoutAndRedirect()}
        testID='logout'>
          <Text style={styles.userButtonText}>Logout</Text>
        </TouchableOpacity>
      </View>
    </View>
  )
}

function mapStateToProps (state: any) {
  return {
    user: state.userReducer.user
  }
}

function mapDispatchToProps (dispatch: any) {
  return {
    action: bindActionCreators({ userUpdate, userLogout }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(User)
