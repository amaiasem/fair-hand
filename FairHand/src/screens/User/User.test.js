import React from 'react'
import { Provider } from 'react-redux'
import { cleanup, render, fireEvent } from '@testing-library/react-native'
import configureStore from 'redux-mock-store'
import User from './User'
import mockUser from '../../../constants/mockUser.ts'

const mockStore = configureStore()

describe('Given a component User', () => {
  let store
  let component
  let navigate

  beforeEach(() => {
    store = mockStore({
      userReducer: { user: mockUser }
    })
    navigate = jest.fn()
    component = (
      <Provider store={store}>
        <User navigation={{ navigate }}/>
        </Provider>)
  })

  afterEach(() => cleanup())

  it('It should match the snapshot', () => {
    render(component)
    expect(component).toMatchSnapshot()
  })
})

describe('When pressing on MyReviews button', () => {
  let store
  let component
  let navigate

  beforeEach(() => {
    store = mockStore({
      userReducer: {
        user: {
          _id: '604f3dcafcd75136b496ec65',
          name: 'Amaia Semper',
          surname: '',
          email: 'amaia@gmail.com',
          password: 'ahg56',
          address: 'Travessera de les Corts, Barcelona',
          myFavourites: [
            {
              shopId: '6046ab3fb9f8122550ebebc9'
            }
          ]
        }
      }
    })
    navigate = jest.fn()
    component = (
    <Provider store={store}>
      <User navigation={{ navigate }}/>
      </Provider>)
  })

  it('It should call navigation.navigate with argument MyReviews', () => {
    const { getByTestId } = render(component)

    fireEvent.press(getByTestId('navigate-myreviews'))
    expect(navigate).toHaveBeenCalledWith('MyReviews')
  })
})
