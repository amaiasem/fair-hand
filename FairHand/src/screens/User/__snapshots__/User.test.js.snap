// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a component User Then should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 30,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "elevation": 4,
        "flexDirection": "row",
        "height": 50,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.23,
        "shadowRadius": 2.62,
        "width": 750,
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 40,
          "marginTop": 3,
          "width": 150,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.5,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://instagram.fqsa1-1.fna.fbcdn.net/v/t51.2885-15/e35/s1080x1080/60793471_340043599979046_7591053955213517299_n.jpg?tp=1&_nc_ht=instagram.fqsa1-1.fna.fbcdn.net&_nc_cat=106&_nc_ohc=RbfXgsjm6WgAX_p8uD1&oh=653befd6c914a74146eedf0ff34134cd&oe=607936F6",
        }
      }
      style={
        Object {
          "borderRadius": 80,
          "height": 150,
          "margin": 20,
          "width": 150,
        }
      }
    />
    <Text
      style={
        Object {
          "fontSize": 26,
          "fontWeight": "600",
        }
      }
    >
      Amaia Semper
    </Text>
    <Text
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "600",
        }
      }
    >
      amaia@gmail.com
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.4,
        "justifyContent": "flex-end",
      }
    }
  >
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FF7F50",
            "borderRadius": 25,
            "height": 35,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "width": 637.5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 18,
              "fontWeight": "700",
            }
          }
        >
          My favourites
        </Text>
      </View>
    </RNGestureHandlerButton>
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FF7F50",
            "borderRadius": 25,
            "height": 35,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "width": 637.5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 18,
              "fontWeight": "700",
            }
          }
        >
          My reviews
        </Text>
      </View>
    </RNGestureHandlerButton>
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
    >
      <View
        accessible={true}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#262322",
            "borderRadius": 25,
            "height": 35,
            "justifyContent": "center",
            "marginTop": 10,
            "opacity": 1,
            "width": 637.5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fff",
              "fontSize": 18,
              "fontWeight": "700",
            }
          }
        >
          Logout
        </Text>
      </View>
    </RNGestureHandlerButton>
  </View>
</View>
`;
